### PR TITLE
Create facilities to simplify setting function and parameter attributes

### PR DESCRIPTION
**Pull request by etiotto**

### CODE REVIEW

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp [hunk 1]
```diff
@@ -49,16 +49,39 @@ using namespace mlir::triton::gpu;
 // Helper Functions
 //===----------------------------------------------------------------------===//
 
-static intel::AttributeList
-getAttrList(const intel::AttrBuilder &funcAttrBuilder,
-            ArrayRef<NamedAttrList> paramAttrs = {}) {
+static void addToAttrList(intel::AttributeList &functionAttrs,
+                          llvm::ArrayRef<NamedAttrList> paramAttrs) {
+  functionAttrs.addParamAttributes(paramAttrs);
+}
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:55]: Is this function used?

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp [hunk 2]
```diff
@@ -627,9 +625,13 @@ createGenISA2DBlockWrite(TritonGEN::Matrix2DBlockStoreOp op,
                           cache,
                           storeVal};
 
-  auto callOp = rewriter.create<LLVM::CallOp>(loc, funcOp, args);
-  callOp.setCConv(LLVM::cconv::CConv::SPIR_FUNC);
-  return callOp;
+  intel::AttributeList attrs =
+      createFunctionAttributes({{llvm::Attribute::NoUnwind, std::nullopt},
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:629]: FYI by the GenISA definition, `GenISA_LSC2DBlockWrite` doesn't have any attributes.

**Review by chengjunlu — APPROVED** [review]

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp [hunk 3]
```diff
@@ -49,16 +49,39 @@ using namespace mlir::triton::gpu;
...
-            ArrayRef<NamedAttrList> paramAttrs = {}) {
+static void addToAttrList(intel::AttributeList &functionAttrs,
+                          llvm::ArrayRef<NamedAttrList> paramAttrs) {
+  functionAttrs.addParamAttributes(paramAttrs);
+}
+
+intel::AttributeList createFunctionAttributes(
+    const ArrayRef<
+        std::pair<llvm::Attribute::AttrKind, std::optional<uint64_t>>>
+        attributes,
```
Comment (victor-eds) [inline third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:60]:
```suggestion
    ArrayRef<
        std::pair<llvm::Attribute::AttrKind, std::optional<uint64_t>>>
        attributes,
```
This is usually passed this like this

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp [hunk 4]
```diff
@@ -49,16 +49,39 @@ using namespace mlir::triton::gpu;
...
+
   intel::AttributeList attrs;
   attrs.addFnAttributes(funcAttrBuilder);
-  if (!paramAttrs.empty())
-    attrs.addParamAttributes(paramAttrs);
   return attrs;
 }
 
+NamedAttrList
+createParameterAttributes(const ArrayRef<llvm::Attribute::AttrKind> attributes,
```
Comment (victor-eds) [inline third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:76]: Same as above

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp [hunk 5]
```diff
@@ -49,16 +49,39 @@ using namespace mlir::triton::gpu;
...
+        std::pair<llvm::Attribute::AttrKind, std::optional<uint64_t>>>
+        attributes,
+    MLIRContext *ctx) {
+  intel::AttrBuilder funcAttrBuilder(*ctx);
+  for (auto &attr : attributes) {
+    if (attr.second)
+      funcAttrBuilder.addPassthroughAttribute(attr.first, *attr.second);
+    else
+      funcAttrBuilder.addPassthroughAttribute(attr.first);
+  }
```
Comment (victor-eds) [inline third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:68]:
```suggestion
  for (auto [kind, optValue] : attributes) {
    if (optValue)
      funcAttrBuilder.addPassthroughAttribute(kind, optValue);
    else
      funcAttrBuilder.addPassthroughAttribute(kind);
  }
```
A bit cleaner IMO

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp (hunk 1 above)
Comment (etiotto) [reply third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:55]: removed, it was used at some point but I inlined it and forgot to remove the function definition.

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp (hunk 3 above)
Comment (etiotto) [reply third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:60]: ok... removed `const`

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp (hunk 5 above)
Comment (etiotto) [reply third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:68]: done

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp (hunk 4 above)
Comment (etiotto) [reply third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:76]: done

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp (hunk 2 above)
Comment (etiotto) [reply third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:629]: Still, it should not throw an exception.

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp [hunk 6]
```diff
@@ -1288,11 +1242,14 @@ struct TritonSubGroupShuffleLowering
                   ConversionPatternRewriter &rewriter) const override {
     Value val = op.getValue();
     auto origTy = val.getType();
-    val = extend(op, op.getValue(), origTy, rewriter);
+
+    origTy.print(llvm::errs());
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:1246]: remove

**Review by whitneywhtsang — APPROVED** [review]

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp (hunk 6 above)
Comment (etiotto) [reply third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:1246]: done

**Pull request merged**